### PR TITLE
Don't auto-instantiate a DRBG when trying to use it and it's not

### PR DESCRIPTION
**Pull request by kroeckx**
The one creating the DRBG should instantiate it, it's there that we
know which parameters we should use to instantiate it.

This splits the rand init in two parts to avoid a deadlock
because when the global drbg is created it wands to call
rand_add on the global rand method.

### CODE REVIEW

File: crypto/rand/drbg_lib.c [hunk 1]
```diff
@@ -397,14 +439,31 @@ static int drbg_status(void)
...
-        RAND_DRBG_instantiate(&rand_drbg, NULL, 0);
     ret = rand_drbg.state == DRBG_READY ? 1 : 0;
     CRYPTO_THREAD_unlock(rand_drbg.lock);
     return ret;
 }
 
+RAND_DRBG *RAND_DRBG_get0_global(void)
+{
+    if (!RUN_ONCE(&rand_init_drbg, do_rand_init_drbg))
+        return 0;
```
Comment (richsalz) [inline crypto/rand/drbg_lib.c]: NULL

File: crypto/rand/drbg_lib.c [hunk 2]
```diff
@@ -397,14 +439,31 @@ static int drbg_status(void)
...
+
+    return &rand_drbg;
+}
+
+RAND_DRBG *RAND_DRBG_get0_priv_global(void)
+{
+    if (!RUN_ONCE(&rand_init_drbg, do_rand_init_drbg))
+        return 0;
```
Comment (richsalz) [inline crypto/rand/drbg_lib.c]: NULL

File: crypto/rand/rand_lib.c [hunk 3]
```diff
@@ -365,15 +334,13 @@ void RAND_add(const void *buf, int num, double randomness)
...
 
     if (meth != RAND_OpenSSL())
         return RAND_bytes(buf, num);
 
-    if (priv_drbg.state == DRBG_UNINITIALISED
-            && RAND_DRBG_instantiate(&priv_drbg, NULL, 0) == 0)
-        return 0;
-    return RAND_DRBG_generate(&priv_drbg, buf, num, 0, NULL, 0);
-
+    drbg = RAND_DRBG_get0_priv_global();
```
Comment (richsalz) [inline crypto/rand/rand_lib.c]: check for NULL?

File: crypto/rand/drbg_lib.c [hunk 4]
```diff
@@ -397,14 +439,31 @@ static int drbg_status(void)
 {
     int ret;
 
+    if (!RUN_ONCE(&rand_init_drbg, do_rand_init_drbg))
```
Comment (richsalz) [inline crypto/rand/drbg_lib.c]: Have this call the get0-rand function

File: crypto/rand/drbg_lib.c [hunk 5]
```diff
@@ -340,14 +330,66 @@ void *RAND_DRBG_get_ex_data(const RAND_DRBG *drbg, int idx)
...
+    free_drbg(&rand_drbg);
+    free_drbg(&priv_drbg);
+}
+
 static int drbg_bytes(unsigned char *out, int count)
 {
     int ret = 0;
     size_t chunk;
 
+    if (!RUN_ONCE(&rand_init_drbg, do_rand_init_drbg))
```
Comment (richsalz) [inline crypto/rand/drbg_lib.c]: have this call the get0 rand function

**Comment by kroeckx** [comment]
I think I addressed all you review comments

**Comment by kroeckx** [comment]
There were conflicts because randomness was removed, please re-review.

**Pull request closed without merging**